File: services/memory-service/app/api/routes/tasks.py
"""任务与待办端点（M-10，docs/01 §6.6）。

  GET    /api/tasks            列表（status 过滤，updated_at 倒序）
  POST   /api/tasks            创建（可选 query session_id 自动挂链）
  PATCH  /api/tasks/{task_id}  更新（PATCH 语义；状态变更同步进度记忆）

Agent（todo 工具）与前端共用同一 TaskService；鉴权：Bearer JWT +
workspace 成员校验（CurrentWorkspaceQuery，403 防枚举）。
"""

import uuid as uuid_mod
from typing import Annotated

from fastapi import APIRouter, Query

from app.core.deps import CurrentUser, CurrentWorkspaceQuery, DbDep
from app.models.enums import TaskStatus
from app.schemas.task import TaskCreate, TaskRead, TaskUpdate
from app.services.task_service import task_service

router = APIRouter(prefix="/api/tasks", tags=["tasks"])


@router.get("", response_model=list[TaskRead], summary="任务列表")
async def list_tasks(
    ws: CurrentWorkspaceQuery,
    db: DbDep,
    status: Annotated[TaskStatus | None, Query(description="状态过滤；缺省为全部")] = None,
) -> list[TaskRead]:
    """workspace 内任务清单（最近变动在前）。"""
    tasks = await task_service.list_tasks(db, ws_id=ws.id, status=status)
    return [TaskRead.model_validate(task) for task in tasks]


@router.post("", status_code=201, response_model=TaskRead, summary="创建任务")
async def create_task(
    ws: CurrentWorkspaceQuery,
    user: CurrentUser,
    db: DbDep,
    payload: TaskCreate,
    session_id: Annotated[
        uuid_mod.UUID | None, Query(description="来源会话（自动挂入关联列表）")
    ] = None,
) -> TaskRead:
    """创建任务并同步 job.*.progress 记忆（Task Continuity 起点）。"""
    task = await task_service.create(
        db, ws_id=ws.id, user_id=user.id, payload=payload, session_id=session_id
    )
    await db.commit()
    await db.refresh(task)
    return TaskRead.model_validate(task)


@router.patch("/{task_id}", response_model=TaskRead, summary="更新任务")
async def update_task(
    task_id: uuid_mod.UUID,
    ws: CurrentWorkspaceQuery,
    user: CurrentUser,
    db: DbDep,
    payload: TaskUpdate,
    session_id: Annotated[
        uuid_mod.UUID | None, Query(description="来源会话（自动挂入关联列表）")
    ] = None,
) -> TaskRead:
    """更新任务（仅提交字段生效；状态/标题变更触发记忆同步）。

    Raises:
        NotFoundError: 任务不存在（404）。
    """
    task = await task_service.update(
        db, ws_id=ws.id, task_id=task_id, user_id=user.id, payload=payload, session_id=session_id
    )
    await db.commit()
    await db.refresh(task)
    return TaskRead.model_validate(task)

File: frontend/src/api/tasks.ts
/**
 * 任务/待办 API（EchoDesk 前端，M-F5）。
 *
 * 对接 memory-service `/api/tasks`（M-10）：与 Agent 的 todo 工具共用
 * 同一 TaskService，状态变更自动同步 job.*.progress 记忆。
 * workspace_id 走 query 参数；创建/更新可选挂 source 会话。
 */
import { api } from "./client";

/** 任务状态（服务端 TaskStatus）。 */
export type TaskStatus = "open" | "doing" | "done" | "abandoned";

/** 任务条目。 */
export interface Task {
  id: string;
  title: string;
  status: TaskStatus;
  /** 0 普通 / 1 高（开场注入简报）。 */
  priority: number;
  due_date: string | null;
  related_session_ids: string[];
  created_at: string;
  updated_at: string;
}

/** 创建任务请求体。 */
export interface TaskCreatePayload {
  title: string;
  priority?: number;
  due_date?: string | null;
}

/** 更新任务请求体（PATCH 语义：仅提交字段生效）。 */
export interface TaskUpdatePayload {
  title?: string;
  status?: TaskStatus;
  priority?: number;
  due_date?: string | null;
}

/**
 * 任务列表（最近变动在前）。
 *
 * @param workspaceId - workspace ID（query 参数）。
 * @param status - 状态过滤；缺省全部。
 * @returns 任务列表。
 */
export async function listTasks(
  workspaceId: string,
  status?: TaskStatus
): Promise<Task[]> {
  const { data } = await api.get<Task[]>("/api/tasks", {
    params: { workspace_id: workspaceId, ...(status ? { status } : {}) },
  });
  return data;
}

/**
 * 创建任务（并同步 job.*.progress 记忆）。
 *
 * @param workspaceId - workspace ID。
 * @param payload - 标题/优先级/截止时间。
 * @returns 新建任务。
 */
export async function createTask(
  workspaceId: string,
  payload: TaskCreatePayload
): Promise<Task> {
  const { data } = await api.post<Task>("/api/tasks", payload, {
    params: { workspace_id: workspaceId },
  });
  return data;
}

/**
 * 更新任务（PATCH 语义；状态/标题变更触发记忆同步）。
 *
 * @param workspaceId - workspace ID。
 * @param taskId - 任务 ID。
 * @param patch - 待更新字段。
 * @returns 更新后的任务。
 */
export async function updateTask(
  workspaceId: string,
  taskId: string,
  patch: TaskUpdatePayload
): Promise<Task> {
  const { data } = await api.patch<Task>(`/api/tasks/${taskId}`, patch, {
    params: { workspace_id: workspaceId },
  });
  return data;
}
